FIX MLPClassifier not iterating to max_iter if warm started

norbusan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Fixes #16812

#### What does this implement/fix? Explain your changes.

MLPClassifier's `_fit` routine uses
```
 incremental = (self.warm_start and hasattr(self, "classes_"))
```
which, in the case of warm_start and a previous fit call evaluates to
true, thus the fit routine breaks after one iteration instead of
continuing to `max_iter`.
    
The solution requires three changes:
- `MLPClassifier.fit` calls `self._fit` with `incremental = False`
  (similar to MLPRegressor)
- Input validation needs to be updated so that the check for
  compatibility of classes with previous calls is done when `warm_start`
  is true, and either incremental is True or previously created
  results from fit are available in `classes_`
- internal iteration counter needs to be reset to 0 if incremental
  is False, but `warm_start` is True.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

norbusan committed:
Fix warm_start behaviour for MLPClassifier (Fixes #16812)

MLPClassifier's _fit routine uses
	incremental = (self.warm_start and hasattr(self, "classes_"))
which, in the case of warm_start and a previous fit call evaluates to
true, thus the fit routine breaks after one iteration instead of
continuing to max_iter.

The solution requires three changes:
- MLPClassifier.fit calls self._fit with incremental = False
  (similar to MLPRegressor)
- Input validation needs to be updated so that the check for
  compatibility of classes with previous calls is done when warm_start
  is true, and either incremental is True or previously created
  results from fit are available in classes_
- internal iteration counter needs to be reset to 0 if incremental
  is False, but warm_start is True.

norbusan committed:
Add a test for Issue 16812: warm starting MLPClassifier

glemaitre commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -662,6 +662,17 @@ def test_warm_start():
             clf.fit(X, y_i)
 
 
+def test_warm_start_full_iteration():
```
We should parametrize the test with pytest and check that we have the same behaviour with the regressor and classifier

```suggestion
@pytest.mark.parametrize("MLPEstimator", [MLPClassifier, MLPRegressor])
def test_warm_start_full_iteration(MLPEstimator):
    # Non-regression test for:
    # https://github.com/scikit-learn/scikit-learn/issues/16812
    # Check that the MLP estimator only accomplish `max_iter` with a
    # warm started estimator.
```

norbusan replied:
Good point, didn't think of that. I will look into it and update the PR (but leaving for a weekend of mountaineering soon)

norbusan replied:
Done that with the last set of commits pushed

glemaitre commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -662,6 +662,17 @@ def test_warm_start():
+    max_iter = 3
+    clf = MLPClassifier(hidden_layer_sizes=2, solver='sgd',
+                        warm_start=True, max_iter=max_iter)
+    clf.fit(X, y)
```
```suggestion
    clf.fit(X, y)
    assert max_iter == clf.n_iter_
```

norbusan replied:
That is a bit hard to read since it is the same code, do you mean something in the vein of
```
clf.fit(X, y)
assert max_iter == clf.n_iter_
clf.fit(X, y)
assert max_iter == clf.n_iter_
```
That is, testing two times that `max_iter` has been reached?

norbusan replied:
Done that with the last set of commits pushed.

glemaitre commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -1037,8 +1043,7 @@ def fit(self, X, y):
         """
-        return self._fit(X, y, incremental=(self.warm_start and
-                                            hasattr(self, "classes_")))
+        return self._fit(X, y, incremental=False)
```
We can actually remove the `fit` function because it is the same as the base class then.

norbusan replied:
Indeed indeed, good point! I will update the PR.

Do you prefer a push --force with the initial commit removing the function, a a commit on top of the current status?

thomasjpfan replied:
We prefer to commit on top because it helps us keep the context of the comments on github.

norbusan replied:
Thanks @thomasjpfan
I'll send new commits after the weekend (currently sweating the hell out of it climbing).

norbusan replied:
Done that with the last set of commits pushed.

glemaitre reviewed (commented):
A quick first pass. I will make a more thorough review in the afternoon.

norbusan commented:
Thanks @glemaitre for the initial review and useful comments. Looking forward to your in-detailed review.

norbusan committed:
drop MLPClassifier::fit as it is the same as super()::fit

norbusan committed:
parametrize test_warm_start_full_iteration

test both MLPEstimators for complete iteration up to max_iter
under warm_start.

Suggestion by Guillaume Lemaitre

norbusan committed:
test_warm_start_full_iteration: check max_iter after first fit, too

rth commented on sklearn/neural_network/_multilayer_perceptron.py:
```diff
@@ -977,17 +982,18 @@ def _validate_input(self, X, y, incremental):
+        if not incremental:
+            self._label_binarizer = LabelBinarizer()
+            self._label_binarizer.fit(y)
+            self.classes_ = self._label_binarizer.classes_
```
The issue though is that with a warm start (where now `incremental=False`), the `self.classes_` should not be redefined, unless it was not set previously.

norbusan replied:
Indeed, thanks for the detailed check. I pushed a new commit the reorders the checks in the validate function, and also add a matrix of actions taken to make it easier to see why what is done where.

glemaitre replied:
I don't know if this is possible, but we should have had a test failing.

norbusan replied:
I'm not sure now (away from kbd) but I think this test was actually dead code in the previous version, that is the reason it couldn't fail in a test. I guess with the current setup it should be possible to trigger a failure.

norbusan replied:
To be more specific than my previous comment, now that I am back at the kbd:

Before, there was no way that `incremental == True` and `warm_start == False`, because `incremental` was passed in as `self.warm_start and hasattr(self, "classes_")`.

The "dead code" I mentioned above is actually not dead code, but can happen when `MLPClassifier::partial_fit` is called with incorrect initial classes, then this is already tested in `test_mlp.py::test_partial_fit_classes_error`.

norbusan committed:
Fix incorrect redefinition of classes_

in case that incremental == False and warm_start == True, the current
status would redefine classes_ even if they are already present.

Rework the test logic to only redefine in cases where it is meant to
be (not incremental, not warm_start, or classes_ not defined)

norbusan committed:
fix line length of re-indented code

instead of doing this in the previous commit, this is done separately
to allow for easier checking using git diff -w

norbusan commented:
One comment concerning the last commit: I wanted to silent the flake8 warnings of excessive line length, so I reformatted the messages that extended the limit due to the re-indentation. I have done this in a separate commit so that checks via `git diff -w` give reasonable results.

glemaitre commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -662,6 +662,23 @@ def test_warm_start():
+    # https://github.com/scikit-learn/scikit-learn/issues/16812
+    # Check that the MLP estimator accomplish `max_iter` with a
+    # warm started estimator.
+    X = X_iris
+    y = y_iris
```
```suggestion
    X, y = X_iris, y_iris
```

glemaitre commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -662,6 +662,23 @@ def test_warm_start():
+    X = X_iris
+    y = y_iris
+    max_iter = 3
+    clf = MLPEstimator(hidden_layer_sizes=2, solver='sgd',
+                       warm_start=True, max_iter=max_iter)
```
```suggestion
    clf = MLPEstimator(
        hidden_layer_sizes=2, solver='sgd', warm_start=True, max_iter=max_iter
    )
```

glemaitre reviewed (approved):
Only some nitpick and one place that it would be more explicit to add `self.warm_start`

norbusan commented:
Thanks @glemaitre for the approval and "nit-picking". Are you fine with adding the changes directly via the github interface (which will have you as creator I guess)? Other option is that I add commit(s) for the changes.

glemaitre commented:
As you want :) We are going to squash all commits at the end to have a
single commit so it does not really matter.

On Wed, 2 Sep 2020 at 10:17, Norbert Preining <notifications@github.com>
wrote:

> Thanks @glemaitre <https://github.com/glemaitre> for the approval and
> "nit-picking". Are you fine with adding the changes directly via the github
> interface (which will have you as creator I guess)? Other option is that I
> add commit(s) for the changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/18269#issuecomment-685435463>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABY32P255MJX7WIHE56BVV3SDX5RJANCNFSM4QMSGXQQ>
> .
>


-- 
Guillaume Lemaitre
Scikit-learn @ Inria Foundation
https://glemaitre.github.io/

norbusan committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

norbusan commented:
Thanks for the quick answer, done that now.

glemaitre commented:
@rth @thomasjpfan do you want to give it a fresh look

thomasjpfan commented on sklearn/neural_network/tests/test_mlp.py:
```diff
@@ -662,6 +662,23 @@ def test_warm_start():
+    clf.fit(X, y)
+    assert max_iter == clf.n_iter_
+    clf.fit(X, y)
+    assert max_iter == clf.n_iter_
```
The use case of warm_start is to allow for the increase of `max_iter` to add more iterations:

```py
clf.fit(X, y)
assert clf.n_iter_ == 3

# keeps the first 3 iterations and trains for 4 more iterations.
clf.set_params(max_iter=7)
clf.fit(X, y)
assert clf.n_iter_ == 7
```

In the case of `warm_start=True` and an unchanged `max_iter`, I would suspect the second call to `fit` would be a noop.

norbusan replied:
Thanks for your comment. I'm (again) away from kbd in the mountains, but AFAIR this is not the case. If warm start is true the starting classes are initialized with the outcome of the last fit. That would mean that the second call to fit here with three iterations will still give different results. 

But please correct me if I'm wrong, not having the code at hand now.

norbusan replied:
@thomasjpfan thanks for your comment. I see how this makes sense in random forests, but I don't see how - or what - we could change/improve here at the moment.

We can change the number of max iters, but that will just give another different model, since in total the model went not through 3+3, but through 3+n iterations. While this can be made a test, I don't see any difference in the **check** we can make, since there is nothing similar to the `n_estimators`.

To spell it differently: in my restricted knowledge, it seems that warm starting without changing the number of estimators will not change the outcome in the case of random forests, but as far as I see it will change the outcome int he case of mlpclassifiers - or am I wrong here? That said, I am not really sure what kind of improvement you would like to see here, besides giving different number of max iterations for the second call?

thomasjpfan replied:
`n_estimators` and `max_iter` are the parameters that are warm started for their respective estimators. But the implementation of warm_start in MLP is different. It can warm start without changing the parameters, which needs to be discussed outside of this PR.

That being said, the `n_iter_` behavior for `MLPRegressor` is correct:

### master

```py
from sklearn.neural_network import MLPRegressor
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
X, y = make_regression(n_samples=200, random_state=1)
regr = MLPRegressor(max_iter=10, warm_start=True).fit(X, y)

assert regr.n_iter_ == 10

regr.fit(X, y)
assert regr.n_iter_ == 20
```

### This PR

```py
from sklearn.neural_network import MLPRegressor
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
X, y = make_regression(n_samples=200, random_state=1)
regr = MLPRegressor(max_iter=10, warm_start=True).fit(X, y)

assert regr.n_iter_ == 10

regr.fit(X, y)
assert regr.n_iter_ == 10
```

From the original issue: https://github.com/scikit-learn/scikit-learn/issues/16812 we want to make sure that `MLPClassifier` also follows the behavior of `MLPRegressor` + `n_iter_`.

> That said, I am not really sure what kind of improvement you would like to see here, besides giving different number of max iterations for the second call?

Yes, preserving the behavior on master regarding `n_iter_` is what I am suggesting.

norbusan replied:
Thanks for your patience, now I understand the issue, and I am aware of it.

That was in fact a conscious change by myself, because I consider the original behavior (shown above in master branch) to increase the number of iterations done when `incremental == False` as incorrect.

I assumed - probably wrongly - that `incremental == False` should mean we start training from the beginning (`n_iter_ = 0`) for the full set of max iterations. 

Orthogonal to this, if `warm_start == True`, then the starting model is not random, but the outcome from a previous fit.

But these two are orthogonal, meaning, that if `warm_start == True` and `incremental == False` we start at iteration 0, etc.

Furthermore, the only distinguishable difference is not in the actual outcome, but only in the value of `n_iter_`, other than this there are no behavioral differences.

I guess the bottomline is the following question of mine: What is the expected/correct behavior of `n_iter` under `incremental == False` when calling `fit`?

norbusan replied:
To be more specific: My initial proposal contained this code
```
--- a/sklearn/neural_network/_multilayer_perceptron.py
+++ b/sklearn/neural_network/_multilayer_perceptron.py
@@ -379,6 +379,11 @@ class BaseMultilayerPerceptron(BaseEstimator, metaclass=ABCMeta):
                                            incremental):
             # First time training the model
             self._initialize(y, layer_units, X.dtype)
+        # In case we are warm starting but aren't incremental, the
+        # above routine is not called and the n_iter_ counter is not reset,
+        # do it here.
+        if not incremental:
+            self.n_iter_ = 0
```

related to my above statements.

Reverting the current state of this PR to the behavior of master is easy with these changes:
```
diff --git a/sklearn/neural_network/_multilayer_perceptron.py b/sklearn/neural_network/_multilayer_perceptron.py
index bdab6b49f..2937e59e3 100644
--- a/sklearn/neural_network/_multilayer_perceptron.py
+++ b/sklearn/neural_network/_multilayer_perceptron.py
@@ -379,10 +379,6 @@ class BaseMultilayerPerceptron(BaseEstimator, metaclass=ABCMeta):
                                            incremental):
             # First time training the model
             self._initialize(y, layer_units, X.dtype)
-        if self.warm_start and not incremental:
-            # only the number of iteration should be initialized since
-            # weights are already allocated due to the warm start.
-            self.n_iter_ = 0
 
         # Initialize lists
         activations = [X] + [None] * (len(layer_units) - 1)
diff --git a/sklearn/neural_network/tests/test_mlp.py b/sklearn/neural_network/tests/test_mlp.py
index f8ff68445..72f347966 100644
--- a/sklearn/neural_network/tests/test_mlp.py
+++ b/sklearn/neural_network/tests/test_mlp.py
@@ -676,7 +676,7 @@ def test_warm_start_full_iteration(MLPEstimator):
     clf.fit(X, y)
     assert max_iter == clf.n_iter_
     clf.fit(X, y)
-    assert max_iter == clf.n_iter_
+    assert 2 * max_iter == clf.n_iter_
 
 
 def test_n_iter_no_change():
```

I am happily pushing this to the PR, please let me know.

norbusan replied:
I think the change of behavior of `n_iter_` shouldn't be indeed part of the PR, and if we want to change this behavior, it should be done in a separate PR.

Thus, I have pushed the change and reverted to the behavior as in master with respect to `n_iter_`.

Thanks for you detailed considerations!

thomasjpfan reviewed (commented):
Thank you for the PR @norbusan !

norbusan committed:
Revert behavior wrt n_iter_ to the state in master

thomasjpfan reviewed (approved):
Please add an entry to the change log at `doc/whats_new/v0.24.rst` with tag |Fix|. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

norbusan committed:
update news document

norbusan commented:
Thanks for the approval @thomasjpfan 
I have now pushed an entry to the news file.

norbusan commented:
@thomasjpfan thanks for merging
